File: server/src/billing.ts
import { createHmac, timingSafeEqual } from 'node:crypto';
import type { Express, Request, Response } from 'express';
import Razorpay from 'razorpay';
import type { Store } from './store.js';

/**
 * Razorpay subscriptions.
 *
 * Bought on the web rather than through Play Billing, because the app ships to more than the Play
 * Store, because a desktop build cannot use Play Billing at all, and because one entitlement
 * service is simpler than one-and-a-half. The trade is that we own the subscription lifecycle:
 * renewals, failed payments, cancellations. That is what the webhook below is.
 *
 * The app never opens any of this. It signs in; the buying happens on the web, in a browser,
 * where store policy has nothing to say about it.
 */

export const PLAN_ID = process.env.RAZORPAY_PLAN_ID ?? '';

function client(): Razorpay | null {
  const key_id = process.env.RAZORPAY_KEY_ID;
  const key_secret = process.env.RAZORPAY_KEY_SECRET;
  if (!key_id || !key_secret) return null;
  return new Razorpay({ key_id, key_secret });
}

/** HMAC-SHA256 of the RAW body, compared in constant time. */
export function webhookSignatureMatches(body: Buffer, signature: string, secret: string): boolean {
  const expected = createHmac('sha256', secret).update(body).digest('hex');
  const a = Buffer.from(expected, 'utf8');
  const b = Buffer.from(signature, 'utf8');
  return a.length === b.length && timingSafeEqual(a, b);
}

/**
 * Map a Razorpay subscription status onto ours.
 *
 * Razorpay has more states than the entitlement question needs. `halted` means retries have been
 * exhausted, which is a harder stop than `pending`; both are past_due to us, and the difference
 * that matters — how long the grace lasts — is decided by the period end, not the label.
 */
function mapStatus(status: string): 'active' | 'past_due' | 'cancelled' | 'none' {
  switch (status) {
    case 'active':
    case 'authenticated':
      return 'active';
    case 'pending':
    case 'halted':
      return 'past_due';
    case 'cancelled':
    case 'completed':
    case 'expired':
      return 'cancelled';
    default:
      return 'none';
  }
}

export function billingRoutes(app: Express, store: Store): void {
  /**
   * Start a subscription. Returns Razorpay's hosted checkout URL to redirect to.
   *
   * The hosted page rather than an embedded checkout: it handles UPI AutoPay, cards, netbanking
   * and their respective failure paths, all of which we would otherwise be reimplementing and
   * getting subtly wrong on somebody's bank.
   */
  app.post('/api/billing/subscribe', async (req: Request, res: Response) => {
    const rzp = client();
    if (!rzp || !PLAN_ID) {
      return res.status(503).json({
        error: 'Billing is not configured on this server (RAZORPAY_KEY_ID, RAZORPAY_KEY_SECRET, RAZORPAY_PLAN_ID).',
      });
    }
    const { email, password } = req.body ?? {};
    if (typeof email !== 'string' || typeof password !== 'string') {
      return res.status(400).json({ error: 'Sign in to subscribe' });
    }
    const account = store.authenticate(email, password);
    if (!account) return res.status(401).json({ error: 'Email or password is incorrect' });

    try {
      const subscription = await rzp.subscriptions.create({
        plan_id: PLAN_ID,
        // 120 monthly cycles — Razorpay requires a finite count, and ten years is long enough
        // that nobody reaches it before we have changed something else.
        total_count: 120,
        customer_notify: 1,
        notes: { accountId: account.id, email: account.email },
      });

      // Recorded as pending BEFORE the user pays, so the webhook that follows can find the
      // account from the subscription id. A payment whose account we cannot identify is a
      // support ticket and a refund.
      store.upsertSubscription({
        accountId: account.id,
        plan: 'pro',
        providerId: subscription.id,
        status: 'none',
        currentPeriodEnd: 0,
      });

      res.json({ subscriptionId: subscription.id, checkoutUrl: (subscription as { short_url?: string }).short_url });
    } catch (e) {
      const message = e instanceof Error ? e.message : 'Could not start the subscription';
      res.status(502).json({ error: message });
    }
  });

  /**
   * Razorpay's webhook. The only thing that may mark a subscription paid.
   *
   * Never trust a redirect back from a checkout page to mean payment succeeded — a browser can be
   * pointed anywhere by anyone. The webhook is signed, and it is what moves money into
   * entitlement.
   */
  app.post('/api/billing/webhook', (req: Request, res: Response) => {
    const secret = process.env.RAZORPAY_WEBHOOK_SECRET;
    if (!secret) return res.status(503).send('webhook not configured');

    const signature = req.header('x-razorpay-signature') ?? '';
    const raw = req.body as Buffer;
    if (!Buffer.isBuffer(raw) || !webhookSignatureMatches(raw, signature, secret)) {
      return res.status(400).send('bad signature');
    }

    let event: {
      event?: string;
      payload?: { subscription?: { entity?: Record<string, unknown> } };
    };
    try {
      event = JSON.parse(raw.toString('utf8'));
    } catch {
      return res.status(400).send('bad json');
    }

    // Razorpay retries on any non-2xx, so the same charge can arrive twice. Claiming the event id
    // first means a retry is acknowledged without extending the period a second time.
    const eventId = req.header('x-razorpay-event-id') ?? '';
    if (eventId && !store.claimEvent(eventId)) return res.status(200).send('duplicate');

    const entity = event.payload?.subscription?.entity;
    if (!entity) return res.status(200).send('ignored');

    const providerId = String(entity.id ?? '');
    const accountId = providerId ? store.accountByProviderId(providerId) : null;
    if (!accountId) {
      // 200, not an error: retrying will not make an unknown subscription known, and a webhook
      // Razorpay keeps retrying forever is noise that hides real failures.
      return res.status(200).send('unknown subscription');
    }

    const status = mapStatus(String(entity.status ?? ''));
    const periodEnd = Number(entity.current_end ?? entity.end_at ?? 0) || 0;
    const existing = store.subscription(accountId);

    store.upsertSubscription({
      accountId,
      plan: 'pro',
      providerId,
      status,
      // Never move the paid-through date backwards. Events can arrive out of order, and the
      // version that does move it back takes away time somebody has already paid for.
      currentPeriodEnd: Math.max(existing.currentPeriodEnd, periodEnd),
    });

    res.status(200).send('ok');
  });
}

File: server/app/billing.py
"""
Becoming entitled.

Purchases happen through Google Play Billing, inside the app, and nowhere else. That is not a
simplification imposed on the design — it is the shape of the business: one store, one payment
provider, one way to become a subscriber.

The client cannot be trusted with entitlement. A purchase token is a string and anyone can send a
string, so every one of them is checked against Google and only Google's answer is written down.
There is no path in this module by which a request can make itself paid.

Cancellation is here too, because it is the other half of the same guarantee: deleting an account
must stop the money as well as the record.
"""

from __future__ import annotations

from dataclasses import dataclass

from . import play
from .store import Store, Subscription


def cancel_play_subscription(provider_id: str) -> tuple[bool, str | None]:
    """
    Cancel a subscription at Google, immediately.

    Used when an account is deleted, and it is the step that must not be skipped: deleting our row
    stops us knowing about a subscription, it does not stop Google charging the card. Somebody
    billed monthly for an account they deleted is the worst failure this system can produce, so the
    caller refuses to delete when this returns False.

    Two answers count as success, because in both of them there is no card left to stop:

    * The subscription is already cancelled or expired — it is in the state we wanted, and calling
      that a failure would strand the account forever, unable to be deleted.
    * Google has never heard of the token (404). A forged or long-dead purchase token cannot be
      cancelled and does not need to be.

    Everything else — a timeout, a 5xx, a revoked service account — is a *maybe*, and a maybe must
    refuse the deletion. The asymmetry is deliberate: refusing costs somebody a second attempt,
    while wrongly succeeding costs them money every month.
    """
    if not play.is_configured():
        return False, "Billing is not configured on this server."
    try:
        purchase = play.verify(provider_id)
    except play.PlayError as e:
        if e.status == 404:
            return True, None
        return False, str(e)

    if purchase.status == "cancelled":
        return True, None
    if not purchase.product_id:
        # Live, but Google returned no line items to name the product with. The URL cannot be
        # built, so the honest answer is that we could not cancel it.
        return False, "Google did not say which product this subscription is for."

    try:
        play.cancel(provider_id, purchase.product_id)
    except play.PlayError as e:
        if e.status == 404:
            return True, None
        return False, str(e)
    return True, None


@dataclass(frozen=True)
class LinkResult:
    status: int
    account_id: str | None = None
    error: str | None = None


def link_play_purchase(store: Store, purchase_token: str) -> LinkResult:
    """
    Turn a Play purchase token into an entitled account, creating one if this is a first purchase.

    Order matters throughout:

    **Verify before anything is written.** A token Google has never heard of must not create an
    account, or the table fills with rows conjured by anyone who can POST.

    **Follow linkedPurchaseToken.** When somebody upgrades a plan or resubscribes after lapsing,
    Google issues a NEW token and points it at the old one. Missing that link would strand the
    original account — the customer keeps their devices and their history on an account that no
    longer has a subscription, while a fresh empty account holds the thing they paid for.

    **Acknowledge last.** Google auto-refunds anything unacknowledged within three days, so it must
    happen — but only after entitlement is safely recorded. Acknowledging first and then failing to
    write would leave somebody paying for nothing, with Google satisfied they had been served.
    """
    if not play.is_configured():
        return LinkResult(status=503, error="Play Billing is not configured on this server.")
    if not purchase_token:
        return LinkResult(status=400, error="purchaseToken is required")

    try:
        purchase = play.verify(purchase_token)
    except play.PlayError as e:
        # 502, not 400: as far as we know the token may be perfectly good and Google unreachable.
        # The app retries, and nothing has been granted in the meantime.
        return LinkResult(status=502, error=str(e))

    if purchase.status == "none":
        # PENDING — a payment method still being authorised, which happens with some UPI mandates
        # and cash-based methods. Not an error, just not yet paid.
        return LinkResult(status=402, error="That purchase has not completed yet.")

    account_id = store.account_by_provider_id(purchase_token, provider="play")
    if account_id is None and purchase.linked_purchase_token:
        account_id = store.account_by_provider_id(purchase.linked_purchase_token, provider="play")
    if account_id is None:
        # No email, no password: Google has established who this is, and making somebody invent a
        # password to receive what they have just paid for is friction that buys nothing.
        account_id = store.create_account().id

    store.upsert_subscription(
        Subscription(
            account_id=account_id,
            plan="pro",
            provider="play",
            provider_id=purchase_token,
            status=purchase.status,
            # Deliberately not clamped to never move backwards: Google's answer IS the
            # truth here, it is fetched fresh rather than pushed, and it cannot arrive out of order.
            current_period_end=purchase.expires_at,
        )
    )

    if not purchase.acknowledged:
        try:
            play.acknowledge(purchase_token, purchase.product_id)
        except play.PlayError:
            # Entitlement is already recorded, so the customer has what they paid for. A failed
            # acknowledgement is retried on their next licence refresh, well inside Google's three
            # days — and refusing here would take away access over a bookkeeping call.
            pass

    return LinkResult(status=200, account_id=account_id)


def refresh_play_subscription(store: Store, account_id: str) -> None:
    """
    Re-ask Google what a Play subscription is worth, and record the answer.

    Called on every licence refresh, which is what lets cancellations and expiries propagate with
    no real-time notifications, no Pub/Sub topic and no service to keep running. The app refreshes
    about weekly and a token lasts a fortnight, so the worst case is a cancelled subscriber keeping
    Pro slightly longer — which is the direction to err in.

    Silent on failure. A Google outage must not revoke a paying customer, and their existing token
    keeps working until it expires either way.
    """
    sub = store.subscription(account_id)
    if sub.provider != "play" or not sub.provider_id:
        return
    try:
        purchase = play.verify(sub.provider_id)
    except play.PlayError:
        return

    store.upsert_subscription(
        Subscription(
            account_id=account_id, plan="pro", provider="play",
            provider_id=sub.provider_id, status=purchase.status,
            current_period_end=purchase.expires_at,
        )
    )
    if not purchase.acknowledged:
        try:
            play.acknowledge(sub.provider_id, purchase.product_id)
        except play.PlayError:
            pass
